lang/cpp.go

```diff
@@ -0,0 +1,633 @@
+	case '\r': is.ignore(); goto L2;
+	case '\n': is.ignore(); goto L3;
+	case EOF : goto L3;
+	}
+L3: int n = dq.size();
+	for (int i = 0, j = 1; i < n; ++i) {
+		auto root = dq[i];
+		if (root == nullptr) { continue; }
+		root->left = j < n ? dq[j] : nullptr;
+		root->right = j + 1 < n ? dq[j + 1] : nullptr;
+		j += 2;
+	}
+	node = n ? dq[0] : nullptr;
+	return is;
+}`,
+	"auto &operator<<(ostream &os, TreeNode *node)": `{
+	queue<TreeNode *> q;
+	int cnt_not_null_nodes = 0;
+	auto push = [&](TreeNode *node) {
+		q.emplace(node);
+		if (node != nullptr) { ++cnt_not_null_nodes; }
+	};
+	auto pop = [&]() {
+		auto front = q.front(); q.pop();
+		if (front != nullptr) {
+			--cnt_not_null_nodes;
+			push(front->left);
+			push(front->right);
+			os << front->val << ',';
+		} else {
+			os << "null,";
+		}
+	};
+	os << '[';
+	if (node != nullptr) {
+		push(node);
+		while (cnt_not_null_nodes > 0) { pop();	}
+		os.seekp(-1, ios_base::end);
+	}
+	os << ']';
+	return os;
+}`,
+	`template <typename T>
+auto &operator>>(istream &is, vector<T> &v)`: `{
+L0:	is.ignore();
+L1: switch (is.peek()) {
+	case ' ':
+	case ',': is.ignore(); goto L1;
+	case ']': is.ignore(); goto L2;
+	default : v.emplace_back();
+	          if constexpr (is_same_v<T, string>) {
+	              ` + cppGen.GetScanCodeForType(0, "string", "v.back()", "is") + `
+	          } else {
+	              is >> v.back();
+	          }
+	          goto L1;
+	}
+L2: switch (is.peek()) {
+	case '\r': is.ignore(); goto L2;
+	case '\n': is.ignore(); goto L3;
+	case EOF : goto L3;
+}
+L3: return is;
+}`,
+	`template <typename T>
+auto &operator<<(ostream &os, const vector<T> &v)`: `{
+	os << '[';
+	if constexpr (is_same_v<T, string>) {
+		for (auto &&x : v) { ` + cppGen.GetPrintCodeForType(0, "string", "x", "os") + ` os << ','; }
+	} else if constexpr (is_same_v<T, double>) {
+		for (auto &&x : v) { ` + cppGen.GetPrintCodeForType(0, "double", "x", "os") + ` os << ','; }
+	} else {
+		for (auto &&x : v) { os << x << ','; }
+	}
+	os.seekp(-1, ios_base::end);
+	os << ']';
+	return os;
+}`,
+}
+
+var cppTypes = map[string]string{
+	"integer":  "int",
+	"string":   "string",
+	"double":   "double",
+	"ListNode": "ListNode*",
+	"TreeNode": "TreeNode*",
+}
+
+const (
+	objectName           = "obj"
+	returnName           = "res"
+	inputFileStreamName  = "ifs"
+	outputFileStreamName = "ofs"
+	cppHeader            = `#include <bits/stdc++.h>
+using namespace std;
+
+// definitions
+%s
+
+#include "solution.h"
+
+// helper funcs
+%s
+
+// main func
+int main(int argc, char **argv) {
+	if (argc != 3) {
+		return 1;
+	}
+
+	ifstream ifs(argv[1]);
+	if (!ifs.is_open()) {
+		return 1;
+	}
+
+	ofstream ofs(argv[2]);
+	if (!ofs.is_open()) {
+		return 1;
+	}
+
+	// scan input args
+%s
+
+	// initialize object
+%s
+
+	// call methods
+%s
+
+	// print result
+%s
+
+	// delete object
+	delete ` + objectName + `;
+
+	ifs.close();
+	ofs.close();
+	return 0;
+}
+`
+)
+
+func (c cpp) GetCppType(t string) (int, string) {
```

Comment:
Internal functions should not be upper-cased.

Reply:
fixed